routers/api/packages/arch/arch.go

```diff
@@ -0,0 +1,204 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package arch
+
+import (
+	"bytes"
+	"encoding/hex"
+	"io"
+	"net/http"
+	"strings"
+
+	pkg_model "code.gitea.io/gitea/models/packages"
+	"code.gitea.io/gitea/modules/context"
+	pkg_module "code.gitea.io/gitea/modules/packages"
+	arch_module "code.gitea.io/gitea/modules/packages/arch"
+	"code.gitea.io/gitea/routers/api/packages/helper"
+	pkg_service "code.gitea.io/gitea/services/packages"
+	arch_service "code.gitea.io/gitea/services/packages/arch"
+)
+
+// Push new package to arch package registry.
+func Push(ctx *context.Context) {
+	var (
+		owner    = ctx.Params("username")
+		filename = ctx.Params("filename")
+		distro   = ctx.Params("distro")
+		sign     = ctx.Params("sign")
+	)
+
+	upload, close, err := ctx.UploadStream()
+	if err != nil {
+		apiError(ctx, http.StatusInternalServerError, err)
+		return
+	}
+	if close {
+		defer upload.Close()
+	}
+
+	buf, err := pkg_module.CreateHashedBufferFromReader(upload)
+	if err != nil {
+		apiError(ctx, http.StatusInternalServerError, err)
+		return
+	}
+	defer buf.Close()
+
+	desc, err := arch_module.EjectMetadata(&arch_module.EjectParams{
+		Filename:     filename,
+		Distribution: distro,
+		Buffer:       buf,
+	})
+	if err != nil {
+		apiError(ctx, http.StatusBadRequest, err)
+		return
+	}
+
+	_, err = buf.Seek(0, io.SeekStart)
+	if err != nil {
+		apiError(ctx, http.StatusInternalServerError, err)
+		return
+	}
+
+	props := map[string]string{
+		distro + "-" + filename + ".desc": desc.String(),
+	}
+	if sign != "" {
+		_, err := hex.DecodeString(sign)
+		if err != nil {
+			apiError(ctx, http.StatusBadRequest, err)
+		}
+		props[distro+"-"+filename+".sig"] = sign
+	}
+
+	ver, _, err := pkg_service.CreatePackageOrAddFileToExisting(
+		&pkg_service.PackageCreationInfo{
+			PackageInfo: pkg_service.PackageInfo{
+				Owner:       ctx.Package.Owner,
+				PackageType: pkg_model.TypeArch,
+				Name:        desc.Name,
+				Version:     desc.Version,
+			},
+			Creator: ctx.ContextUser,
+			Metadata: &arch_module.Metadata{
+				URL:          desc.URL,
+				Description:  desc.Description,
+				Provides:     desc.Provides,
+				License:      desc.License,
+				Depends:      desc.Depends,
+				OptDepends:   desc.OptDepends,
+				MakeDepends:  desc.MakeDepends,
+				CheckDepends: desc.CheckDepends,
+				Backup:       desc.Backup,
+			},
+		},
+		&pkg_service.PackageFileCreationInfo{
+			PackageFileInfo: pkg_service.PackageFileInfo{
+				Filename:     filename,
+				CompositeKey: distro + "-" + filename,
```

Comment:
Adding the filename does not change anything or?

Reply:
To ensure that composite key is unique for different distributions for the same file name.

Reply:
Fixed, removed unnecessary filename from composite key.